testing/src/main/java/org/oppia/android/testing/network/BUILD.bazel

```diff
@@ -41,4 +41,18 @@ kt_android_library(
     ],
 )
 
+kt_android_library(
+    name = "retrofit_module",
+    srcs = [
+        "RetrofitModule.kt",
```

Comment:
Is this correct? This doesn't seem to be corresponding to a real module unless I'm missing something. Is this supposed to be ``RetrofitTestModule.kt``?

Reply:
Thanks Ben for pointing this out. 

I listed the following deps for this library -
```starlark
kt_android_library(
    name = "retrofit_module",
    srcs = [
        "RetrofitTestModule.kt",
    ],
    visibility = ["//:oppia_testing_visibility"],
    deps = [
        ":dagger",
        ":prod_module",
        "//third_party:com_squareup_retrofit2_converter-moshi",
    ],
)
```

But I am getting this error - 
![Screenshot from 2021-04-28 11-56-19](https://user-images.githubusercontent.com/54636525/116356660-30203d00-a819-11eb-8821-63b08dc1435d.png)

Then I tried removing the prod_module dependency - 
```starlark
kt_android_library(
    name = "retrofit_module",
    srcs = [
        "RetrofitTestModule.kt",
    ],
    visibility = ["//:oppia_testing_visibility"],
    deps = [
        ":dagger",
        "//third_party:com_squareup_retrofit2_converter-moshi",
    ],
)
```

And got this error-  
```
ERROR: /home/prayutsu/StudioProjects/oppia-android/testing/src/main/java/org/oppia/android/testing/network/BUILD.bazel:44:19: KotlinCompile //testing/src/main/java/org/oppia/android/testing/network:retrofit_module_kt { kt: 1, java: 0, srcjars: 0 } for k8 failed: (Exit 1): builder failed: error executing command bazel-out/host/bin/external/io_bazel_rules_kotlin/src/main/kotlin/builder ... (remaining 1 argument(s) skipped)
error: /tmp/working9325588898131114075/_kotlinc/testing_src_main_java_org_oppia_android_testing_network-retrofit_module_kt_jvm/temp/stubs/org/oppia/android/testing/network/RetrofitTestModule.java:18: error: incompatible types: NonExistentClass cannot be converted to Annotation
    @error.NonExistentClass()
          ^
testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:5:26: error: unresolved reference: data
import org.oppia.android.data.backends.gae.OppiaRetrofit
                         ^
testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:7:18: error: unresolved reference: mock
import retrofit2.mock.MockRetrofit
                 ^
testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:8:18: error: unresolved reference: mock
import retrofit2.mock.NetworkBehavior
                 ^
testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:20:28: error: unresolved reference: OppiaRetrofit
  fun provideMockRetrofit(@OppiaRetrofit retrofit: Retrofit): MockRetrofit {
                           ^
testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:20:63: error: unresolved reference: MockRetrofit
  fun provideMockRetrofit(@OppiaRetrofit retrofit: Retrofit): MockRetrofit {
                                                              ^
Compilation failure: compile phase failed:
error: /tmp/working9325588898131114075/_kotlinc/testing_src_main_java_org_oppia_android_testing_network-retrofit_module_kt_jvm/temp/stubs/org/oppia/android/testing/network/RetrofitTestModule.java:18: error: incompatible types: NonExistentClass cannot be converted to Annotation
    @error.NonExistentClass()
          ^
/home/prayutsu/.cache/bazel/_bazel_prayutsu/ddf5f2ba1270603694fd5d626215d411/execroot/__main__/testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:5:26: error: unresolved reference: data
import org.oppia.android.data.backends.gae.OppiaRetrofit
                         ^
/home/prayutsu/.cache/bazel/_bazel_prayutsu/ddf5f2ba1270603694fd5d626215d411/execroot/__main__/testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:7:18: error: unresolved reference: mock
import retrofit2.mock.MockRetrofit
                 ^
/home/prayutsu/.cache/bazel/_bazel_prayutsu/ddf5f2ba1270603694fd5d626215d411/execroot/__main__/testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:8:18: error: unresolved reference: mock
import retrofit2.mock.NetworkBehavior
                 ^
/home/prayutsu/.cache/bazel/_bazel_prayutsu/ddf5f2ba1270603694fd5d626215d411/execroot/__main__/testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:20:28: error: unresolved reference: OppiaRetrofit
  fun provideMockRetrofit(@OppiaRetrofit retrofit: Retrofit): MockRetrofit {
                           ^
/home/prayutsu/.cache/bazel/_bazel_prayutsu/ddf5f2ba1270603694fd5d626215d411/execroot/__main__/testing/src/main/java/org/oppia/android/testing/network/RetrofitTestModule.kt:20:63: error: unresolved reference: MockRetrofit
  fun provideMockRetrofit(@OppiaRetrofit retrofit: Retrofit): MockRetrofit {
                                                              ^
INFO: Elapsed time: 5.146s, Critical Path: 4.82s
INFO: 11 processes: 8 internal, 3 worker.
FAILED: Build did NOT complete successfully
//data:src/test/java/org/oppia/android/data/backends/gae/JsonPrefixNetworkInterceptorTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/ClassroomServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/ConceptCardServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/ExplorationServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/FeedbackReportingServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/QuestionPlayerServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/StoryServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/SubtopicServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/backends/gae/api/TopicServiceTest NO STATUS
//data:src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest NO STATUS

FAILED: Build did NOT complete successfully

```
I am unable to fix this, I also tried changing the visibility of the libraries. Can you please help here?

Reply:
I have fixed all the build errors but the tests are failing locally for me on Bazel - 
```starlark
 Test output for //data:src/test/java/org/oppia/android/data/backends/gae/api/QuestionPlayerServiceTest:
JUnit4 Test Runner
.E
Time: 1.088
There was 1 failure:
1) testQuestionPlayerService_usingFakeJson_deserializationSuccessful(org.oppia.android.data.backends.gae.api.QuestionPlayerServiceTest)
java.lang.NoClassDefFoundError: android/util/TypedValue
        at java.base/java.lang.Class.getDeclaredFields0(Native Method)
        at java.base/java.lang.Class.privateGetDeclaredFields(Class.java:3220)
        at java.base/java.lang.Class.getDeclaredField(Class.java:2567)
        at org.robolectric.shadows.ShadowAssetManager$_AssetManager28_$$Reflector0.setSystemApkAssetsSet(Unknown Source)
        at org.robolectric.shadows.ShadowArscAssetManager9.reset(ShadowArscAssetManager9.java:233)
        at org.robolectric.Shadows.reset(Shadows.java:2304)
        at org.robolectric.android.internal.AndroidTestEnvironment.resetState(AndroidTestEnvironment.java:518)
        at org.robolectric.RobolectricTestRunner.lambda$finallyAfterTest$0(RobolectricTestRunner.java:349)
        at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:75)
        at org.robolectric.RobolectricTestRunner.finallyAfterTest(RobolectricTestRunner.java:347)
        at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:272)
        at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.ClassNotFoundException: couldn't load android.util.TypedValue
        at org.robolectric.internal.bytecode.SandboxClassLoader.maybeInstrumentClass(SandboxClassLoader.java:147)
        at org.robolectric.internal.bytecode.SandboxClassLoader.lambda$loadClass$0(SandboxClassLoader.java:115)
        at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:53)
        at org.robolectric.internal.bytecode.SandboxClassLoader.loadClass(SandboxClassLoader.java:115)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 16 more
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 59
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:195)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:176)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:162)
        at org.robolectric.internal.bytecode.ClassNodeProvider.createClassNode(ClassNodeProvider.java:24)
        at org.robolectric.internal.bytecode.ClassNodeProvider.getClassNode(ClassNodeProvider.java:16)
        at org.robolectric.internal.bytecode.InstrumentingClassWriter.typeInfo(InstrumentingClassWriter.java:115)
        at org.robolectric.internal.bytecode.InstrumentingClassWriter.getCommonSuperClass(InstrumentingClassWriter.java:36)
        at org.objectweb.asm.SymbolTable.addMergedType(SymbolTable.java:1200)
        at org.objectweb.asm.Frame.merge(Frame.java:1299)
        at org.objectweb.asm.Frame.merge(Frame.java:1244)
        at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1610)
        at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1546)
        at org.objectweb.asm.MethodVisitor.visitMaxs(MethodVisitor.java:764)
        at org.objectweb.asm.tree.MethodNode.accept(MethodNode.java:767)
        at org.objectweb.asm.tree.MethodNode.accept(MethodNode.java:647)
        at org.objectweb.asm.tree.ClassNode.accept(ClassNode.java:408)
        at org.robolectric.internal.bytecode.ClassInstrumentor.instrumentToBytes(ClassInstrumentor.java:80)
        at org.robolectric.internal.bytecode.SandboxClassLoader.lambda$maybeInstrumentClass$2(SandboxClassLoader.java:139)
        at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:53)
        at org.robolectric.internal.bytecode.SandboxClassLoader.maybeInstrumentClass(SandboxClassLoader.java:138)
        ... 20 more

FAILURES!!!
Tests run: 1,  Failures: 1
```

Looks like there's some problem with Robolectric version - https://github.com/robolectric/robolectric/issues/5082

Reply:
Ah. Do you have JDK 8 installed locally? Bazel uses an embedded JDK 11 for actions & building, but for running it uses whatever version of JDK you have locally (we generally recommend Java 8, though CI uses 9 I believe). This seems to be a class version incompatibility, so installing JDK 8 & setting it as your default may fix this issue.

If you don't want to set it as your default, you can set JAVA_HOME temporarily before using ``bazel test`` and it should use that version. There might be a way to configure this via bazelrc, as well, but I'm not sure.

Reply:
I think I earlier had JDK 8 installed but I recently removed it, but I will configure it again to run bazel tests locally whenever required.
Thanks @BenHenning.

Reply:
No problem. Just to check though, did that fix the issue @prayutsu?

Reply:
Yup, it fixed the issue :)